src/net_processing.cpp

```diff
@@ -2566,6 +2572,33 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
         return;
     }
 
+    if (msg_type == NetMsgType::BLOCKRELAY) {
+        if (pfrom.fSuccessfullyConnected) {
+            // Disconnect peers that send blockrelay message after VERACK; this
+            // must be negotiated between VERSION and VERACK.
+            pfrom.fDisconnect = true;
+            return;
+        }
+        if (pfrom.IsFullOutboundConn() || pfrom.IsAddrFetchConn()) {
+            // If we picked an outbound for tx- or addr-relay and it sends us a
+            // BLOCKRELAY, find another peer.
+            pfrom.fDisconnect = true;
+            return;
+        }
+        if (pfrom.m_tx_relay != nullptr && WITH_LOCK(pfrom.m_tx_relay->cs_filter, return pfrom.m_tx_relay->fRelayTxes)) {
+            // Can't send a BLOCKRELAY message if they didn't turn off tx-relay
+            // in the version message.
+            pfrom.fDisconnect = true;
+        }
+        if (pfrom.IsInboundConn()) {
```

Comment:
I thiiink this could be problematic if the peer sends two `BLOCKRELAY` messages in a row?

- initially `ConnectionType::Inbound` 
- they send first `BLOCKRELAY` message, we hit this code & update to `ConnectionType::INBOUND_BLOCK_RELAY`
- they send another `BLOCKRELAY` message, we hit this code again because `IsInboundConn()` would be true 
- `UpdateConnectionType` is invoked, which asserts the connection type is `ConnectionType::INBOUND`.